### PR TITLE
Add event log sender to the Shell

### PR DESCRIPTION
**Pull request by sug0**
Based on https://github.com/anoma/namada/pull/534

Completes the first task of issue https://github.com/anoma/namada/issues/563

### CODE REVIEW

File: apps/src/lib/node/ledger/mod.rs [hunk 1]
```diff
@@ -442,15 +453,17 @@ fn start_abci_broadcaster_shell(
     // Construct our ABCI application.
     let tendermint_mode = config.tendermint.tendermint_mode.clone();
     let ledger_address = config.shell.ledger_address;
-    let (shell, abci_service) = AbcippShim::new(
+    let event_log_sender = Some(event_log_sender);
```
Comment (batconjurer) [inline apps/src/lib/node/ledger/mod.rs:456]: Shouldn't we pass in a `None` value if we are not a validator?

File: apps/src/lib/node/ledger/shell/mod.rs [hunk 2]
```diff
@@ -854,35 +894,45 @@ mod test_utils {
...
+        /// Returns a new test shell, and a number of channels to mock
+        /// shell sub-processes.
+        ///
+        /// The returned channels include:
+        ///
+        ///    - A channel which will receive any protocol txs sent by the shell
+        ///      via the [`Shell::broadcast`] method.
+        ///    - A channel that can send Ethereum events into the ledger,
+        ///      mocking the Ethereum fullnode process.
         pub fn new_at_height<H: Into<BlockHeight>>(
```
Comment (batconjurer) [inline apps/src/lib/node/ledger/shell/mod.rs:906]: I think we need to pass out the events receiver as well. 1. that way we can test it and 2. if it is dropped, finalize will fail when pushing events to it.

Comment (sug0) [reply apps/src/lib/node/ledger/shell/mod.rs:906]: ah yeah for sure, this is still a TODO item for another task (implementing the event log). I'll update the docstring and return type after that's implemented, this is pretty much just a PR in the spirit of "let's add this thing to the ledger, but we won't actually introduce any new functionality yet"

File: apps/src/lib/node/ledger/mod.rs (hunk 1 above)
Comment (sug0) [reply apps/src/lib/node/ledger/mod.rs:456]: when I implement the event log, I'll update this. this value is pretty much just a placeholder, for the ledger to compile. there's a TODO item above the binding of `event_log_sender` explaining this

File: apps/src/lib/node/ledger/mod.rs [hunk 3]
```diff
@@ -406,7 +408,17 @@ fn start_abci_broadcaster_shell(
...
+    let (broadcast_sender, broadcast_receiver) = mpsc::unbounded_channel();
+
+    // Channels for validators to send events to an event log, which will be
+    // queried from the node's RPC `/events` endpoint.
+    //
+    // TODO: make the sender a param of `start_abci_broadcaster_shell`;
+    // the sender will be returned by another task which starts the
+    // event log; we will also need yet another task for the rpc
+    // server itself. if we are not a validator, `event_log_sender`
+    // should be `None`
```
Comment (james-chf) [inline apps/src/lib/node/ledger/mod.rs:420]: Shouldn't regular full nodes be providing an `/events` endpoint as well? In which case `event_log_sender` shouldn't be an `Option`

**Comment by james-chf** [comment]
Full nodes should provide this `/events` endpoint - if the events are deterministic based on committed blockchain state, they should be able to produce them as well

**Comment by sug0** [comment]
> Full nodes should provide this `/events` endpoint - if the events are deterministic based on committed blockchain state, they should be able to produce them as well

there are also seed nodes, which don't need the RPC endpoint. but you're right, we do need some changes to deal with full nodes.

**Comment by sug0** [comment]
these APIs with optionals will become a lot better once we refactor the shell (https://github.com/anoma/namada/issues/521)

**Review by james-chf — APPROVED** [review]

**Review by batconjurer — APPROVED** [review]

**Pull request merged**